src/meta/src/hummock/compaction/picker/tombstone_reclaim_compaction_picker.rs

```diff
@@ -0,0 +1,135 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use risingwave_pb::hummock::hummock_version::Levels;
+use risingwave_pb::hummock::InputLevel;
+
+use crate::hummock::compaction::overlap_strategy::OverlapStrategy;
+use crate::hummock::compaction::picker::CompactionInput;
+use crate::hummock::level_handler::LevelHandler;
+
+pub struct TombstoneReclaimCompactionPicker {
+    overlap_strategy: Arc<dyn OverlapStrategy>,
+    max_compaction_bytes: u64,
+    delete_ratio_percent: u64,
+    range_delete_ratio_percent: u64,
+}
+
+#[derive(Default)]
+pub struct TombstoneReclaimPickerState {
+    pub last_level: usize,
+}
+
+impl TombstoneReclaimCompactionPicker {
+    pub fn new(
+        overlap_strategy: Arc<dyn OverlapStrategy>,
+        max_compaction_bytes: u64,
+        delete_ratio_percent: u64,
+        range_delete_ratio_percent: u64,
+    ) -> Self {
+        Self {
+            overlap_strategy,
+            max_compaction_bytes,
+            delete_ratio_percent,
+            range_delete_ratio_percent,
+        }
+    }
+
+    pub fn pick_compaction(
+        &self,
+        levels: &Levels,
+        level_handlers: &[LevelHandler],
+        state: &mut TombstoneReclaimPickerState,
+    ) -> Option<CompactionInput> {
+        assert!(!levels.levels.is_empty());
+        let mut select_input_ssts = vec![];
+
+        while state.last_level <= levels.levels.len() {
```

Comment:
default value of `state.last_level` is 0

it might cause panic, since we get the item `state.last_level - 1`

Reply:
fixed